File: packages/stim-cli/src/workspace/settings-schema.ts
import { SETTINGS, type SettingDefinition, type SettingScope } from './settings-registry.ts';

export const SETTINGS_SCHEMA_FILE = 'settings.schema.json';

export const SETTINGS_SCHEMA_URL: string = `https://unpkg.com/stim/dist/${SETTINGS_SCHEMA_FILE}`;

type JsonSchema = Record<string, unknown>;

function leafSchema(setting: SettingDefinition, scope: SettingScope): JsonSchema {
  const { type } = setting;
  const schema: JsonSchema = { description: setting.description };
  switch (type.kind) {
    case 'string':
      schema.type = 'string';
      if (type.pattern) schema.pattern = type.pattern;
      break;
    case 'path':
    case 'bundle-url':
      schema.type = 'string';
      break;
    case 'strings':
      schema.type = 'array';
      schema.items = { type: 'string' };
      break;
    case 'number':
      schema.type = type.integer ? 'integer' : 'number';
      if (type.minimum !== undefined) schema.minimum = type.minimum;
      if (type.maximum !== undefined) schema.maximum = type.maximum;
      break;
    case 'boolean':
      schema.type = 'boolean';
      break;
    case 'choice':
      schema.type = 'string';
      schema.enum = [...type.choices];
      break;
    case 'object':
      schema.type = 'object';
      break;
  }
  if (setting.sensitive && scope === 'committed') schema.pattern = '^(env|file):\\S';
  if (setting.default !== undefined) schema.default = setting.default;
  schema['x-stim'] = {
    key: setting.key,
    kind: type.kind,
    scopes: [...setting.scopes],
    ...(type.kind === 'path' && type.absolute ? { absolute: true } : {}),
    ...(type.kind === 'path' && type.relative ? { relative: true } : {}),
    ...(setting.env ? { env: setting.env } : {}),
    ...(setting.sensitive ? { sensitive: true } : {}),
    ...(setting.committedAt ? { committedAt: setting.committedAt } : {}),
  };
  return schema;
}

function objectSchema(settings: readonly SettingDefinition[], scope: SettingScope, closed: boolean): JsonSchema {
  const root: JsonSchema = { type: 'object', properties: {}, additionalProperties: !closed };
  for (const setting of settings) {
    const parts = setting.key.split('.');
    let node = root;
    for (const part of parts.slice(0, -1)) {
      const properties = node.properties as Record<string, JsonSchema>;
      properties[part] ??= { type: 'object', properties: {}, additionalProperties: false };
      node = properties[part];
    }
    (node.properties as Record<string, JsonSchema>)[parts.at(-1)!] = leafSchema(setting, scope);
  }
  return root;
}

/**
 * The published settings schema. The root describes a committed `.stim.json`;
 * `$defs.machine` describes the settings keys of `$STIM_HOME/config.json`.
 * Every leaf carries `x-stim` with its dotted key, kind, writable scopes,
 * environment override, and sensitivity.
 */
export function settingsJsonSchema(): JsonSchema {
  const committed = objectSchema(
    SETTINGS.filter((setting) => setting.scopes.includes('committed')),
    'committed',
    true,
  );
  (committed.properties as Record<string, JsonSchema>).$schema = { type: 'string' };
  return {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    $id: SETTINGS_SCHEMA_URL,
    title: 'Stim settings',
    description:
      'Settings in a committed .stim.json. $defs.machine lists the machine settings in $STIM_HOME/config.json.',
    ...committed,
    $defs: {
      machine: objectSchema(
        SETTINGS.filter((setting) => setting.scopes.includes('machine')),
        'machine',
        false,
      ),
    },
  };
}

File: apps/desktop/Sources/StimKit/SettingsSchema.swift
import Foundation

public enum SettingScope: String, CaseIterable, Codable, Hashable, Sendable {
  case machine, workspace, repo, committed
}

/// The editor a setting gets, chosen from its schema type.
public enum SettingControl: Hashable, Sendable {
  case picker([String])
  case toggle
  case stepper(minimum: Double?, maximum: Double?, integer: Bool)
  case filePicker
  case tokens
  case json
  case text
  case secure
}

/// One setting described by the `x-stim` annotation of `settings.schema.json`.
public struct SettingField: Hashable, Identifiable, Sendable {
  public var key: String
  public var kind: String
  public var description: String
  public var scopes: [SettingScope]
  public var choices: [String]
  public var minimum: Double?
  public var maximum: Double?
  public var integer: Bool
  public var pattern: String?
  public var defaultValue: JSONValue?
  public var env: String?
  public var sensitive: Bool
  public var committedAtRepository: Bool

  public var id: String { key }

  public var control: SettingControl {
    if sensitive { return .secure }
    switch kind {
    case "choice": return .picker(choices)
    case "boolean": return .toggle
    case "number": return .stepper(minimum: minimum, maximum: maximum, integer: integer)
    case "path": return .filePicker
    case "strings": return .tokens
    case "object": return .json
    default: return .text
    }
  }

  /// The value argument `stim settings set` takes: strings as-is, everything else as JSON.
  public func argument(for value: JSONValue) -> String {
    if case .string(let text) = value { return text }
    return value.json
  }
}

public enum SettingsSchema {
  public static let fileName = "settings.schema.json"

  public struct Invalid: LocalizedError {
    public var errorDescription: String? { "settings.schema.json is not a Stim settings schema." }
  }

  /// Every setting in the schema, each key once: the committed `.stim.json`
  /// tree first, then the machine-only settings from `$defs.machine`, each
  /// tree in dotted-key order.
  public static func fields(from data: Data) throws -> [SettingField] {
    guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw Invalid() }
    var fields: [SettingField] = []
    var seen: Set<String> = []
    let machine = (root["$defs"] as? [String: Any])?["machine"] as? [String: Any]
    for tree in [root, machine].compactMap({ $0 }) {
      collect(tree, into: &fields, seen: &seen)
    }
    guard !fields.isEmpty else { throw Invalid() }
    return fields
  }

  /// `settings.schema.json` beside the resolved `stim` executable, which is how
  /// the npm package ships it (`dist/cli.mjs` and `dist/settings.schema.json`).
  public static func locate(executable: String?, fileManager: FileManager = .default) -> URL? {
    guard let executable else { return nil }
    let resolved = URL(fileURLWithPath: executable).resolvingSymlinksInPath()
    let candidate = resolved.deletingLastPathComponent().appendingPathComponent(fileName)
    return fileManager.fileExists(atPath: candidate.path) ? candidate : nil
  }

  private static func collect(_ node: [String: Any], into fields: inout [SettingField], seen: inout Set<String>) {
    guard let properties = node["properties"] as? [String: Any] else { return }
    for name in properties.keys.sorted() {
      guard let child = properties[name] as? [String: Any] else { continue }
      if let field = field(child), !seen.contains(field.key) {
        seen.insert(field.key)
        fields.append(field)
      } else if child["x-stim"] == nil {
        collect(child, into: &fields, seen: &seen)
      }
    }
  }

  private static func field(_ schema: [String: Any]) -> SettingField? {
    guard let stim = schema["x-stim"] as? [String: Any], let key = stim["key"] as? String,
      let kind = stim["kind"] as? String
    else { return nil }
    let type = schema["type"] as? String
    return SettingField(
      key: key,
      kind: kind,
      description: schema["description"] as? String ?? "",
      scopes: (stim["scopes"] as? [String] ?? []).compactMap(SettingScope.init(rawValue:)),
      choices: schema["enum"] as? [String] ?? [],
      minimum: (schema["minimum"] as? NSNumber)?.doubleValue,
      maximum: (schema["maximum"] as? NSNumber)?.doubleValue,
      integer: type == "integer",
      pattern: schema["pattern"] as? String,
      defaultValue: schema["default"].flatMap(jsonValue),
      env: stim["env"] as? String,
      sensitive: stim["sensitive"] as? Bool ?? false,
      committedAtRepository: stim["committedAt"] as? String == "repository")
  }

  private static func jsonValue(_ any: Any) -> JSONValue? {
    guard JSONSerialization.isValidJSONObject([any]),
      let data = try? JSONSerialization.data(withJSONObject: [any], options: [.fragmentsAllowed])
    else { return nil }
    return (try? JSONDecoder().decode([JSONValue].self, from: data))?.first
  }
}
